### ISSUE TITLE
pubsub sub HTTP command should use text/plain content type

### ISSUE THREAD

**Issue by mitar** [ipfs/go-ipfs#3737]
#### Version information:

```
go-ipfs version: 0.4.5-
Repo version: 5
System version: amd64/darwin
Golang version: go1.7.4
```

#### Description:

Compare:

```
curl -D - "http://localhost:5001/api/v0/log/tail"
HTTP/1.1 200 OK
Access-Control-Allow-Headers: X-Stream-Output, X-Chunked-Output, X-Content-Length
Access-Control-Expose-Headers: X-Stream-Output, X-Chunked-Output, X-Content-Length
Content-Type: text/plain
Server: go-ipfs/0.4.5
Trailer: X-Stream-Error
Vary: Origin
X-Stream-Output: 1
Date: Wed, 01 Mar 2017 10:14:45 GMT
Transfer-Encoding: chunked
```

```
curl -D - "http://localhost:5001/api/v0/pubsub/sub?arg=<topic>&discover=false"
HTTP/1.1 200 OK
Access-Control-Allow-Headers: X-Stream-Output, X-Chunked-Output, X-Content-Length
Access-Control-Expose-Headers: X-Stream-Output, X-Chunked-Output, X-Content-Length
Content-Type: application/json
Server: go-ipfs/0.4.5
Trailer: X-Stream-Error
Vary: Origin
X-Chunked-Output: 1
Date: Wed, 01 Mar 2017 10:14:53 GMT
Transfer-Encoding: chunked
```

Because `pubsub/sub` is content type set to `application/json` code tries to parse it into a JSON. But it is not JSON, but a series of JSON. In fact, trying to use ndjson to parse it fails.

**Comment by mitar** [comment]
Hm, it seems [this is handled specially](https://github.com/ipfs/js-ipfs-api/blob/master/src/request-api.js#L46). In contrast then with log, where [it is parsed here instead](https://github.com/ipfs/js-ipfs-api/blob/master/src/api/log.js#L16).

**Comment by Kubuxu** [comment]
Is the `curl -D - "http://localhost:5001/api/v0/pubsub/sub?arg=<topic>&discover=false"` returning valid ndjson?

**Comment by mitar** [comment]
It is returning valid JSONs per newline. It is just confusing that log returns a different content type than pubsub.

**Comment by Kubuxu** [comment]
We should probably do the reverse then, change the response type of the log tail to json.

**Comment by Kubuxu** [comment]
Really the content type should be `application/x-ndjson`.

**Comment by mitar** [comment]
Yea. Then also [this could be simplified](https://github.com/ipfs/js-ipfs-api/blob/master/src/api/log.js#L16) to simply passing callback to `send`.

**Comment by mitar** [comment]
> Really the content type should be application/x-ndjson.

True, because my browser extension complains now when it tries to parse `application/json`.

**Comment by Kubuxu** [comment]
But it is breaking change to the API, so we have to be really careful with it.

**Comment by Kubuxu** [comment]
@mitar I've edited the title and the main post. Please see if it is correct.

**Comment by mitar** [comment]
Looks good. Thanks!

**Comment by Kubuxu** [comment]
@whyrusleeping @diasdavid @dignifiedquire please take a look at this issue. More we wait for this to happen the harder it will to migrate. I am open to various resolutions. It again reopens problem of API versioning. We claim to be v0 but we treat it as v1.

**Comment by dignifiedquire** [comment]
I'm all for valid content headers.

**Comment by hsanjuan** [comment]
Yeah put valid content headers. This is broken for people that look at that header so I don't think they will complain. And for those that don't look at it it should be harmless. Do we need to update any spec?

**Comment by whyrusleeping** [comment]
Yeah, i'm fine putting in valid content types here.

**Comment by Kubuxu** [comment]
It will break js-ipfs-api, afaik.

**Comment by mitar** [comment]
But a change is simple and mentioned above. Just remove extra ndjson parse.

**Comment by whyrusleeping** [comment]
@Kubuxu eh, unsure. I think they can fix those things to not rely on the content type before we make this change so that it doesnt break when we fix it.

cc @diasdavid @dignifiedquire

**Comment by dignifiedquire** [comment]
Just make the PR, I can fix the api before a release happens.

**Comment by whyrusleeping** [comment]
@dignifiedquire You should be able to fix it so that it works for current behaviour and with the changed content type (making this change on the go side requires no changes to our api client)
